File: user-service/src/main/java/com/study/user/infrastructure/security/TokenRedisService.java
package com.study.user.infrastructure.security;

import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.util.concurrent.TimeUnit;

@Service
@RequiredArgsConstructor
public class TokenRedisService {

    private final StringRedisTemplate redisTemplate;

    private static final String REFRESH_TOKEN_PREFIX = "refresh:";
    private static final String BLACKLIST_PREFIX = "blacklist:";

    public void saveRefreshToken(Long userId, String refreshToken, long expirationMs) {
        redisTemplate.opsForValue().set(
                REFRESH_TOKEN_PREFIX + userId,
                refreshToken,
                expirationMs,
                TimeUnit.MILLISECONDS
        );
    }

    public String getRefreshToken(Long userId) {
        return redisTemplate.opsForValue().get(REFRESH_TOKEN_PREFIX + userId);
    }

    public void deleteRefreshToken(Long userId) {
        redisTemplate.delete(REFRESH_TOKEN_PREFIX + userId);
    }

    public void blacklistAccessToken(String accessToken, long remainingMs) {
        if (remainingMs > 0) {
            redisTemplate.opsForValue().set(
                    BLACKLIST_PREFIX + accessToken,
                    "blacklisted",
                    remainingMs,
                    TimeUnit.MILLISECONDS
            );
        }
    }

    public boolean isBlacklisted(String accessToken) {
        return Boolean.TRUE.equals(redisTemplate.hasKey(BLACKLIST_PREFIX + accessToken));
    }
}

File: user-service/src/main/kotlin/com/study/user/infrastructure/security/TokenRedisService.kt
package com.study.user.infrastructure.security

import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.stereotype.Service
import java.util.concurrent.TimeUnit

@Service
class TokenRedisService(
    private val redisTemplate: StringRedisTemplate
) {
    fun saveRefreshToken(userId: Long, refreshToken: String, expirationMs: Long) {
        redisTemplate.opsForValue().set("refresh:$userId", refreshToken, expirationMs, TimeUnit.MILLISECONDS)
    }

    fun getRefreshToken(userId: Long): String? =
        redisTemplate.opsForValue().get("refresh:$userId")

    fun deleteRefreshToken(userId: Long) {
        redisTemplate.delete("refresh:$userId")
    }

    fun blacklistAccessToken(accessToken: String, remainingMs: Long) {
        if (remainingMs > 0) {
            redisTemplate.opsForValue().set("blacklist:$accessToken", "blacklisted", remainingMs, TimeUnit.MILLISECONDS)
        }
    }

    fun isBlacklisted(accessToken: String): Boolean =
        redisTemplate.hasKey("blacklist:$accessToken") == true
}
